File: csharp/813.largest-sum-of-averages.cs
/*
 * @lc app=leetcode id=813 lang=csharp
 *
 * [813] Largest Sum of Averages
 */

using System;

// @lc code=start
public class Solution {
    public double LargestSumOfAverages(int[] A, int K) {
        int n = A.Length;
        int[] sums = new int[n + 1];
        double[,] dp = new double[n, K];
        for (int i = 0; i < n; ++i) {
            sums[i + 1] = sums[i] + A[i];
        }
        for (int i = 0; i < n; ++i) {
            dp[i, 0] = (double)(sums[n] - sums[i]) / (double)(n - i);
        }
        for (int k = 1; k < K; ++k) {
            for (int i = 0; i < n; ++i) {
                for (int j = i + 1; j < n; ++j) {
                    dp[i, k] = Math.Max(dp[i, k], dp[j, k - 1] + (double)(sums[j] - sums[i]) / (double)(j - i));
                }
            }
        }
        return dp[0, K - 1];
    }
}
// @lc code=end

File: java/813.largest-sum-of-averages.java
/*
 * @lc app=leetcode id=813 lang=java
 *
 * [813] Largest Sum of Averages
 */

// @lc code=start
class Solution {

    private static double largestSumOfAverages(int[] A, int i, int k, Map<String, Double> memo) {
        String key = String.format("%d,%d", i, k);
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        double sum = 0, ret = 0;
        for (int j = i; j + k <= A.length; ++j) {
            sum += (double)A[j];
            if (k > 1) {
                ret = Math.max(ret, sum / (double)(j - i + 1) + largestSumOfAverages(A, j + 1, k - 1, memo));
            }
        }
        if (k == 1) {
            ret = sum / (double)(A.length - i);
        }
        memo.put(key, ret);
        return ret;
    }

    public double largestSumOfAverages(int[] A, int K) {
        return largestSumOfAverages(A, 0, K, new HashMap<String, Double>());
    }
}
// @lc code=end

